django/db/models/indexes.py

```diff
@@ -135,3 +182,61 @@ def __eq__(self, other):
         if self.__class__ == other.__class__:
             return self.deconstruct() == other.deconstruct()
         return NotImplemented
+
+
+class IndexExpression(Func):
+    """
+    Applies correct order and wrapping of expressions suitable for CREATE INDEX statements
+    """
+    template = '%(expressions)s'
+    wrapper_classes = (OrderBy, Collate)
+
+    @classmethod
+    def register_wrappers(cls, *wrapper_classes):
+        cls.wrapper_classes = wrapper_classes
+
+    def resolve_expression(
+        self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False
+    ):
+        expressions = list(self.flatten())
+        index_expressions, wrappers = partition(
+            lambda e: isinstance(e, self.wrapper_classes), expressions
+        )
+        wrapper_types = list(map(type, wrappers))
+        if len(wrapper_types) != len(set(wrapper_types)):
+            raise ValueError(
+                'Multiple references to %s can\'t be used in an indexed expression.' % (
```

Comment:
Minor but pep8 suggested using double quotes instead of escaped single ones

```suggestion
                "Multiple references to %s can't be used in an indexed expression." % (
```

Reply:
TIL, thanks!